Translate current query filters

whikloj opened the pull request:
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2288

# What does this Pull Request do?

If you choose the "Replace PID with Object Label" for a facet, this does not change the filter section of the Islandora Query block.

# What's new?

Change to formatFilter() to check if the "Replace PID with Object Label" setting has been set and if so tries to convert the PID to a label using Solr or loading the object and using the `$object->label`.

# How should this be tested?

Add some facet that displays PIDs, then also add the Islandora Query block (see example screenshots on linked ticket.)

Do a search and the click the converted facet. See that the Filter section of the Islandora Query block is not still converted.

Also the breadcrumbs are wrong.

Pull in this PR. Reload the results page and 

![translate_current_query3](https://user-images.githubusercontent.com/2857697/44039502-d2572e5e-9ede-11e8-8e57-acac1aed3dbd.png)

# Additional Notes:

Example:
* Does this change require documentation to be updated? Unsure
* Does this change add any new dependencies? no
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? no
* Could this change impact execution of existing code? Unsure

# Interested parties
@Islandora/7-x-1-x-committers @lucasvanschaik @janjouketjalsma

whikloj committed:
Translate current query filters

DiegoPino commented on includes/results.inc:
```diff
@@ -550,7 +550,28 @@ class IslandoraSolrResults {
-      $filter_string = $filter_split[1];
+      $facet_fields_settings = islandora_solr_get_fields('facet_fields', TRUE, FALSE, TRUE);
+      $filter_string = str_replace('info:fedora/', '', stripslashes($filter_split[1]));
+      if (isset($facet_fields_settings[$solr_field]['solr_field_settings']['pid_object_label']) && $facet_fields_settings[$solr_field]['solr_field_settings']['pid_object_label'] == 1) {
```
Is this faster than loading the islandora object and getting the label? Just wondering if that single row query could be an overkill?

whikloj replied:
Not sure, I was using the existing translate for the facets as the source for this work. I'm fine just loading the object to get the label.

DiegoPino replied:
Oh, cool. So if this is done like this somewhere else, who am i to say different? 😄 Other committers? I have no strong opinion really

whikloj replied:
Elsewhere the solr query could contain all the facets at once though. So it might be more efficient to just load the object from Fedora to get the label. I'm not sure if a single row Solr query is faster or slower than loading an object from Fedora.

DiegoPino replied:
This piece here https://github.com/Islandora/islandora_solr_search/blob/bcb6301255c297101e43ef6b7110e3362c1a71b2/includes/results.inc#L1001 says lets use islandora_object_load, but could be also based on legacy decisions

DiegoPino commented on includes/results.inc:
```diff
@@ -550,7 +550,28 @@ class IslandoraSolrResults {
+        $qp->solrLimit = 1;
+        $qp->executeQuery(FALSE, TRUE);
+        if ($qp->islandoraSolrResult['response']['numFound'] > 0) {
+          foreach ($qp->islandoraSolrResult['response']['objects'] as $doc) {
```
since solrLimit==1 ,maybe the foreach can omitted and we can get $qp->islandoraSolrResult['response']['objects'][0]['object_label'] if it exists with an (!empty () ?).. not sure if key is 0 there, but you get what i mean.

DiegoPino commented on includes/results.inc:
```diff
@@ -550,7 +550,28 @@ class IslandoraSolrResults {
+            $filter_string = $doc['object_label'];
+          }
+        }
+        else if ($object = islandora_object_load($filter_string)) {
```
Seems like this else if is not of DCS liking.. 
+570: [normal] Use "elseif" in place of "else if"
 +570: [minor] Use "elseif" in place of "else if"

DiegoPino reviewed (dismissed):
It is DCS, not me, i promise.

whikloj committed:
Adjust to use load only

whikloj commented:
_bump_

bondjimbond commented:
@DiegoPino Did you have changes requested besides DCS? We've got one approval on this, but your requested changes (which I think have all been fixed?) are blocking.

DiegoPino commented:
On DLF right now, can only gives this a look tonight

El El lun, 15 de oct. de 2018 a las 11:40, Brandon Weigel <
notifications@github.com> escribió:

> @DiegoPino <https://github.com/DiegoPino> Did you have changes requested
> besides DCS? We've got one approval on this, but your requested changes
> (which I think have all been fixed?) are blocking.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_solr_search/pull/343#issuecomment-429968153>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn857J_sKerlIl54rAkdUe0EklxorOuks5ulNazgaJpZM4V6qez>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

bondjimbond commented:
@DiegoPino One more ping before code freeze. If you have a chance to take a look... does it need more work, or good to go?

DiegoPino commented:
@bondjimbond @whikloj i have two opinions on this. 
A) it works and does what it says 

B) I don't like how its written, sorry @whikloj I really mean well :( Here are the reasons why I don't like it.

1) A [private property](https://github.com/Islandora/islandora_solr_search/pull/343/files#diff-17a5945b1aa412dec3b39453ad29425fR21) without a setter and a [getter](https://github.com/Islandora/islandora_solr_search/pull/343/files#diff-17a5945b1aa412dec3b39453ad29425fR38) that acts as a setter in an invisible way. This makes the whole logic a black box and is probably a not too good OOP practice. 

2) Also, you can't use the class property without having access to `islandora_solr_get_fields('facet_fields', TRUE, FALSE, TRUE);` at all nor set/override it anywhere really.  If it can not be set, and the value is not something that the class itself generates but its source value comes from the outside world, then maybe it is not needed as a property at all?. I know this is Drupal 7 and we kinda mix dependencies/OOP approaches with the fact that we can access any piece of non-OOP code anyway, but if we can avoid this, we should try. 

How do we avoid that? By  (any of this, being the first one the better choice)
 - allowing people to initialize an object of this class (`_construct`) passing `private $facetFieldSettings `as argument, which in this case will contain `islandora_solr_get_fields('facet_fields', TRUE, FALSE, TRUE);`  and by allowing people to set it afterward using a setter function (not obscure anymore). Or 
 - by simply not using a property and accessing the value if needed directly via `islandora_solr_get_fields`. This last option is not 100% oop really but makes the same assumption as current pull, I would not suggest we do it, but what I say is it serves the same purpose as this pull.
- by making the $facetFieldSettings public, and initializing the value when needed, skipping the logic if unset (that means adding a `if !$this->getFacetFieldSettings(); then do the magic` at line 570.

Rest of the logic is Ok and I approve the need and the way.  Maybe there is a rationale on the why this is coded like that and I can be talked out of this. Let me know what do you think. If this is very urgent, well, it does work and I can keep my concerns in check... But if this can be fixed even better.

whikloj commented:
I personally don't use the query filter block so I have no use case for this behaviour nor for needing to change this value. @DiegoPino It sounds like you have some definitive ideas of what you want to see and use, so perhaps you could work on it when you have some time.

br2490 commented:
Patch works as tested for 7.x-1.12. Please see JIRA for images. Uncertain if this helps unblock but perhaps this is a starting point for future development?

DiegoPino commented:
@whikloj @br2490 @bondjimbond it does work, my concerns are on OOp Style and I see @whikloj is Ok with the changes if functionality and his logic stays, so that said, can I make pulls against your branch Jared? If so, I will have to remove my own review, and then someone with less Object-oriented stylish concerns will merge... so kinda egg, chicken and then broken egg situation! Anyone willing to take on the review side of things if i do so?

bondjimbond commented:
@DiegoPino Sure, I'll review when you've added your code.

DiegoPino committed:
Refactor new property facetFieldSettings

- Makes $facetFieldSetting public
- Initializes during _construct
- Resuses a DB call to avoid overhead
- Checks Object access before outputting
- Uses temp variable *($pid) in case object can not be loaded, so
$filter_string  stays same as if [‘pid_object_label'] was disabled.
- DCS *(like 5% of the total reported by me version)

DiegoPino commented:
@whikloj @bondjimbond i added https://github.com/whikloj/islandora_solr_search/pull/1/files

whikloj committed:
Merge pull request #1 from mnylc/translate_current_query-diego

ISLANDORA-2288: Refactor new property facetFieldSettings a bit

DiegoPino committed:
DCS..sorry

DiegoPino commented on includes/results.inc:
```diff
@@ -647,7 +635,9 @@ class IslandoraSolrResults {
   public function prepFieldSubstitutions() {
 
-    $this->facetFieldArray = islandora_solr_get_fields('facet_fields');
+    $facet_fields =  $this->facetFieldSettings = islandora_solr_get_fields('facet_fields', TRUE, FALSE, TRUE);
```
@whikloj i just sent another commit your way, DCS. ... sniff https://github.com/whikloj/islandora_solr_search/pull/2

bondjimbond reviewed (approved):
Tested, works. Approved pending @DiegoPino's DCS.

whikloj committed:
Merge pull request #2 from mnylc/translate_current_query-diego2

DCS..sorry

DiegoPino commented:
@bondjimbond ping. All good with this one?

DiegoPino commented:
@bondjimbond this is a bug fix, @whikloj do we do a release pull?

bondjimbond commented:
@DiegoPino I was going to ask about that. Saw in JIRA it was marked as an "improvement"; didn't know whether we'd consider it a bug or not. But a thing not doing what you'd expect that thing to do feels like a bug. A release pull from @whikloj would be good.

DiegoPino commented:
@bondjimbond i agree. It feels pretty much like a bug. @whikloj what do you think?

whikloj commented:
I kinda think it is an improvement as there is nothing wrong with the way it was operating, but this is better.

bondjimbond commented:
@bryjbrown Thoughts? There's a release PR in queue if you agree.

bryjbrown commented:
@bondjimbond I think I'm leaning towards "bug", since according to this PRs own testing instructions before pulling in the PR the query block and breadcrumbs are wrong and pulling in this PR fixes them. The spirit of the PR is for improvement, but it seems to be correcting erroneous behavior in the process. Especially since there's already a release PR I say go for it :+1:
